### PR TITLE
AMPI: Link with -memory isomalloc and -module CommonLBs by default

### PR DESCRIPTION
**Pull request by evan-charmworks**

### CODE REVIEW

**Comment by matthiasdiener** [comment]
Should this be added to the manual as well?

**Comment by evan-charmworks** [comment]
> Should this be added to the manual as well?

Yes, it should. I will add this to the manual.

**Comment by evan-charmworks** [comment]
Just keeping this up to date. Not ready for review.

File: doc/ampi/manual.rst [hunk 1]
```diff
@@ -348,13 +353,13 @@ adding this directory to the ``$PATH`` environment variable:
 
 These wrappers also allow the user to configure AMPI and Charm++-specific
 functionality.
-To use Isomalloc for transparently migrating user heap data, link with
-*-memory isomalloc*. To use a Charm++ load balancer, link a strategy or
-a suite of strategies in with *-module <LB>*. For example:
+For example, to automatically select a Charm++ load balancer at program
```
Comment (stwhite91) [inline doc/ampi/manual.rst]: We should mention, maybe in a footnote, that AMPI automatically links with CommonLBs

File: src/scripts/configure.ac [hunk 2]
```diff
@@ -109,11 +109,14 @@ remove_make_flag() {
 
 add_flag "OPTS_CC='$OPTS_CC'"   'Pass through user-provided C compiler options'
 add_flag "OPTS_CXX='$OPTS_CXX'" 'Pass through user-provided C++ compiler options'
+add_flag "CMK_WINDOWS='$CMK_WINDOWS'" 'basic setup'
```
Comment (matthiasdiener) [inline src/scripts/configure.ac]: Should be replicated in cmake

File: doc/ampi/manual.rst (hunk 1 above)
Comment (evan-charmworks) [reply doc/ampi/manual.rst]: Done

File: src/scripts/configure.ac (hunk 2 above)
Comment (evan-charmworks) [reply src/scripts/configure.ac]: Done

File: examples/ampi/Cjacobi3D/Makefile [hunk 3]
```diff
@@ -16,10 +21,15 @@ jacobi-cpp: jacobi-cpp.C
 	$(CHARMC) -c jacobi-cpp.C 
 	$(CHARMC) -o jacobi-cpp jacobi-cpp.o -module CommonLBs -lm
 
-jacobi: jacobi.C
+jacobi.o: jacobi.C
 	$(CHARMC) -c jacobi.C 
+
+jacobi: jacobi.o
 	$(CHARMC) -o jacobi jacobi.o -module CommonLBs -lm
```
Comment (stwhite91) [inline examples/ampi/Cjacobi3D/Makefile]: Might as well remove explicit linking with CommonLBs here and elsewhere

**Comment by stwhite91** [comment]
The only other concerns I have with merging this patch before the LB refactor branch are 1) `+LBCommOff` should maybe be on by default, and 2) LBPeriod should maybe be set lower by default. I don't know if there's a performance cost associated with setting LBPeriod lower though? Maybe we could even add a build option to disable periodic LB for now

File: examples/ampi/Cjacobi3D/Makefile (hunk 3 above)
Comment (evan-charmworks) [reply examples/ampi/Cjacobi3D/Makefile]: Done

**Comment by evan-charmworks** [comment]
> The only other concerns I have with merging this patch before the LB refactor branch are 1) `+LBCommOff` should maybe be on by default, and 2) LBPeriod should maybe be set lower by default. I don't know if there's a performance cost associated with setting LBPeriod lower though? Maybe we could even add a build option to disable periodic LB for now

Aren't these issues orthogonal to this PR? In other words, you still have to pass `+balancer` for them to matter?

**Comment by evan-charmworks** [comment]
@matthiasdiener Do you know what's happening with the CMake CI?

```
2020-03-12T22:02:39.1394396Z In file included from /home/runner/work/charm/charm/src/conv-core/threads.C:1707:0:
2020-03-12T22:02:39.1629710Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C: In function â€˜int setJcontext(const uJcontext_t*)â€™:
2020-03-12T22:02:39.1630844Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r4â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1631880Z          : : "r" (startiar), "r" (new_sp), "r" (mu->_uc_args[0]), "r" (mu->_uc_args[1]) : "r1", "r3", "r4", "memory");
2020-03-12T22:02:39.1634634Z                                                                                                                      ^
2020-03-12T22:02:39.1635475Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r3â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1636184Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r1â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1637185Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:157:64: error: unknown register name â€˜r3â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1637938Z          : : "r" (startiar), "r" (mu->uc_link) : "r3", "memory");
2020-03-12T22:02:39.1638695Z                                                                 ^
2020-03-12T22:02:39.1913826Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C: In function â€˜int swapJcontext(uJcontext_t*, const uJcontext_t*)â€™:
2020-03-12T22:02:39.1914683Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r4â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1915322Z          : : "r" (startiar), "r" (new_sp), "r" (mu->_uc_args[0]), "r" (mu->_uc_args[1]) : "r1", "r3", "r4", "memory");
2020-03-12T22:02:39.1915682Z                                                                                                                      ^
2020-03-12T22:02:39.1917175Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r3â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1918621Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r1â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1920704Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:157:64: error: unknown register name â€˜r3â€™ in â€˜asmâ€™
2020-03-12T22:02:39.1921326Z          : : "r" (startiar), "r" (mu->uc_link) : "r3", "memory");
2020-03-12T22:02:39.1921684Z                                                                 ^
2020-03-12T22:02:39.2076162Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C: In function â€˜void CthResume(CthThread)â€™:
2020-03-12T22:02:39.2076979Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r4â€™ in â€˜asmâ€™
2020-03-12T22:02:39.2077632Z          : : "r" (startiar), "r" (new_sp), "r" (mu->_uc_args[0]), "r" (mu->_uc_args[1]) : "r1", "r3", "r4", "memory");
2020-03-12T22:02:39.2078299Z                                                                                                                      ^
2020-03-12T22:02:39.2080021Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r3â€™ in â€˜asmâ€™
2020-03-12T22:02:39.2081521Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:149:117: error: unknown register name â€˜r1â€™ in â€˜asmâ€™
2020-03-12T22:02:39.2084462Z /home/runner/work/charm/charm/src/conv-core/../util/uJcontext.C:157:64: error: unknown register name â€˜r3â€™ in â€˜asmâ€™
2020-03-12T22:02:39.2085060Z          : : "r" (startiar), "r" (mu->uc_link) : "r3", "memory");
2020-03-12T22:02:39.2085189Z                                                                 ^
2020-03-12T22:02:39.2670029Z Fatal Error by charmc in directory /home/runner/work/charm/charm/netlrts-linux-x86_64/src/conv-core
2020-03-12T22:02:39.2671811Z    Command g++ -I/home/runner/work/charm/charm/netlrts-linux-x86_64/bin/../include -D__CHARMC__=1 -DAMPI_ERRHANDLER_RETURN=0 -DCMK_THREADS_ALIAS_STACK=1 -DCMK_THREADS_BUILD_JCONTEXT=1 -I/home/runner/work/charm/charm/src/conv-core/../QuickThreads -I/home/runner/work/charm/charm/netlrts-linux-x86_64/src/QuickThreads/include -I/home/runner/work/charm/charm/src/conv-core/../util -DNDEBUG -O3 -g -U_FORTIFY_SOURCE -std=gnu++11 -O3 -U_FORTIFY_SOURCE -fno-stack-protector -fno-lifetime-dse -std=c++11 -c /home/runner/work/charm/charm/src/conv-core/threads.C -o CMakeFiles/threads-uJcontext-memoryalias.dir/threads.C.o returned error code 1
2020-03-12T22:02:39.2672379Z charmc exiting...
2020-03-12T22:02:39.3379237Z make[2]: *** [src/conv-core/CMakeFiles/threads-uJcontext-memoryalias.dir/threads.C.o] Error 1
```

**Comment by stwhite91** [comment]
Yeah you're right about +balancer

**Comment by evan-charmworks** [comment]
All CIs are fixed, ready for review

**Comment by matthiasdiener** [comment]
> @matthiasdiener Do you know what's happening with the CMake CI?

Just saw this now, this was just a spurious error? Haven't seen it before.

**Comment by evan-charmworks** [comment]
> > @matthiasdiener Do you know what's happening with the CMake CI?
> 
> Just saw this now, this was just a spurious error? Haven't seen it before.

No, it was due to defining `CMK_BLUEGENEQ` as 0 instead of leaving it undefined.

File: examples/ampi/Cjacobi3D/Makefile [hunk 4]
```diff
@@ -7,46 +7,64 @@ TOKENS=6
...
 all: ampi
 
-ampi: jacobi jacobi.iso jacobi.tls jacobi-get
+AMPI_TARGETS := \
+  jacobi \
+  jacobi.pup \
+  jacobi.tls \
+  jacobi-get \
+
+ifeq ($(CMK_WINDOWS)$(CMK_BLUEGENEQ),00)
```
Comment (matthiasdiener) [inline examples/ampi/Cjacobi3D/Makefile]: This is neat, but I'd maybe split them up ;-)

**Pull request merged**